### ISSUE TITLE
Encoding of extension choice (( extensionID: int, extensionValue: any )

### ISSUE THREAD

**Issue by xipki** [cose-wg/CBOR-certificates#171]
Considering version -09, as described in Section 3.1 Message fields (Page 8):

```
 extensions.  The 'extensions' field is encoded as a CBOR array
      where each extension is encoded as either

      -  a CBOR int (see Section 9.4) followed by an optional CBOR item
         of any type, or
     ...
```

If the extensionID is a CBOR int, the extensionValue field may be absent. The application may not be able to parse an unknown extension (this is allowed since the application is allowed to skip non-critical extension), and all subsequent extensions, since it does not know whether the next field is the extensionValue or the extenionID of the next extension.

My suggestion is to change the text to
```
 extensions.  The 'extensions' field is encoded as a CBOR array
      where each extension is encoded as either

      -  a CBOR int (see Section 9.4) followed by an CBOR item
         of any type (null if absent), or
     ...
```

**Comment by gselander** [comment]
```
Extensions = [ * Extension ] / int

Extension = ( extensionID: int, extensionValue: any ) //
            ( extensionID: ~oid, ? critical: true,
              extensionValue: bytes ) //
            ( extensionID: pen, ? critical: true,
              extensionValue: bytes )
```

If I understand right, the issue is to distinguish between an absent extensionValue and the next extension.

If I understand right, the proposed resolution is to require extensionValue = CBOR null instead of absent. This makes sense to me, but we should work on the formulation.

**Comment by xipki** [comment]
> If I understand right, the issue is to distinguish between an absent extensionValue and the next extension.
Yes.

> If I understand right, the proposed resolution is to require extensionValue = CBOR null instead of absent. This makes sense to me, but we should work on the formulation.
Yes.

**Comment by gselander** [comment]
@xipki The CDDL 'any' must be a CBOR-encoded data item, i.e. it cannot be absent. Please give an example when this is a problem to skip an unknown extensionValue.

**Comment by xipki** [comment]
The the text 
```
a CBOR int (see Section 9.4) followed by **an optional** CBOR item of any type, or
```
needs to be changed to
```
a CBOR int (see Section 9.4) followed by *a* CBOR item of any type, or
```

**Comment by gselander** [comment]
Yes, that seems to be an error in the text compared to the CDDL. Changed in #178.

**Comment by emanjon** [comment]
I merged #178 as I see it as a bug fix. Can this be closed?
